okhttp/src/main/java/okhttp3/internal/concurrent/TaskRunner.kt

```diff
@@ -38,117 +37,222 @@ import java.util.concurrent.TimeUnit
+    if (coordinatorWaiting) {
+      backend.coordinatorNotify(this@TaskRunner)
     } else {
-      coordinatorRunning = true
-      backend.executeCoordinator(coordinator)
+      backend.execute(runnable)
     }
   }
 
-  private fun coordinate() {
-    synchronized(this) {
-      while (true) {
-        val now = backend.nanoTime()
-        val delayNanos = executeReadyTasks(now)
+  private fun beforeRun(task: Task) {
+    check(Thread.holdsLock(this))
 
-        if (delayNanos == -1L) {
-          coordinatorRunning = false
-          return
-        }
+    task.nextExecuteNanoTime = -1L
+    val queue = task.queue!!
+    queue.futureTasks.remove(task)
+    readyQueues.remove(queue)
+    queue.activeTask = task
+    busyQueues.add(queue)
+  }
 
-        try {
-          backend.coordinatorWait(this, delayNanos)
-        } catch (_: InterruptedException) {
-          // Will cause the coordinator to exit unless other tasks are scheduled!
-          cancelAll()
-        }
+  private fun runTask(task: Task) {
+    check(!Thread.holdsLock(this))
+
+    val currentThread = Thread.currentThread()
+    val oldName = currentThread.name
+    currentThread.name = task.name
+
+    var delayNanos = -1L
+    try {
+      delayNanos = task.runOnce()
+    } finally {
+      synchronized(this) {
+        afterRun(task, delayNanos)
       }
+      currentThread.name = oldName
+    }
+  }
+
+  private fun afterRun(task: Task, delayNanos: Long) {
+    check(Thread.holdsLock(this))
+
+    val queue = task.queue!!
+    check(queue.activeTask === task)
+
+    val cancelActiveTask = queue.cancelActiveTask
+    queue.cancelActiveTask = false
+    queue.activeTask = null
+    busyQueues.remove(queue)
+
+    if (delayNanos != -1L && !cancelActiveTask && !queue.shutdown) {
+      queue.scheduleAndDecide(task, delayNanos)
+    }
+
+    if (queue.futureTasks.isNotEmpty()) {
+      readyQueues.add(queue)
     }
   }
 
   /**
-   * Start executing the next available tasks for all queues.
-   *
-   * Returns the delay until the next call to this method, -1L for no further calls, or
-   * [Long.MAX_VALUE] to wait indefinitely.
+   * Returns an immediately-executable task for the calling thread to execute, sleeping as necessary
+   * until one is ready. If there are no ready queues, or if other threads have everything under
+   * control this will return null. If there is more than a single task ready to execute immediately
+   * this will launch another thread to handle that work.
    */
-  private fun executeReadyTasks(now: Long): Long {
-    var result = -1L
+  fun awaitTaskToRun(): Task? {
+    check(Thread.holdsLock(this))
+
+    while (true) {
+      if (readyQueues.isEmpty()) {
+        return null // Nothing to do.
+      }
+
+      val now = backend.nanoTime()
+      var minDelayNanos = Long.MAX_VALUE
+      var readyTask: Task? = null
+      var multipleReadyTasks = false
+
+      // Decide what to run. This loop's goal wants to:
```

Comment:
Instead of this loop goal, could you store the next two scheduled times, and loop to find the third at this point?

Reply:
Maybe the question I should answer is, why two items?

The first one is what we’re going to execute on the current thread.
The second one justifies starting a new thread that’ll itself look for up to two items.

We stop after two because once we know we’re going to fork a thread, it becomes _that_ thread’s problem to fork a thread for the third thing.